src/core/util.ts

```diff
@@ -653,20 +653,131 @@ export function isPrimitive(obj: any): boolean {
     return obj[primitiveKey];
 }
 
+interface Iterator<IteratorValue extends unknown, IteratorReturn> {
+    next(): {value: IteratorValue, done: false} | {value: IteratorReturn, done: true};
+}
+
+interface MapInterface<T, KEY extends string | number = string | number> {
+    // Size should be a getter here, but tsconfig will not allow getters (es3 target).
+    // Since our polyfill and HashMap class do not expose this, it should be ok.
+    size(): number;
+    clear(): undefined;
+    delete(key: KEY): boolean;
+    get(key: KEY): T | undefined
+    set(key: KEY, value: T): this
+    has(key: KEY): boolean
+    keys(): Iterator<KEY, undefined>
+    values(): Iterator<T, undefined>
+    entries(): Iterator<[KEY, T], undefined>
+    forEach(callback: (key: KEY, value: T) => void): void
+}
+
+class MapPolyfill<T, KEY extends string | number = string | number> implements MapInterface<T, KEY> {
+    private data: Record<KEY, T> = {} as Record<KEY, T>;
+
+    size(): number {
+        return Object.keys(this.data).length;
+    }
+    toString(): string {
+        return '[object MapPolyfill]';
+    }
+    clear(): undefined {
+        this.data = {} as Record<KEY, T>;
+        return undefined;
+    }
+    delete(key: KEY): boolean {
+        const existed = this.has(key);
+        delete this.data[key];
+        return existed;
+    }
+    get(key: KEY): T | undefined {
+        return this.data[key];
+    }
+    set(key: KEY, value: T): this {
+        this.data[key] = value;
+        return this;
+    }
+    has(key: KEY): boolean {
+        return this.data.hasOwnProperty(key);
+    }
+    keys(): Iterator<KEY, undefined> {
+        const keys = Object.keys(this.data) as KEY[];
+        let i = 0;
+        let end = keys.length;
+        return {
```

Comment:
I think this iterator polyfill can't used in the `for...in..` loop in the browsers don't support iterator because we don't use babel to transpile the code.

We can directly return an array here. And in the `HashMap#keys`, we check if the `data` is an native map and wrap it with `Array.from` if so.

It's a bit dirty but I think it's ok if we don't expose this polyfill and keep it isolated.